File: java/exercises/practice/unit_test_d152de0c_3386_43b4_b0b3_64232c61df7c/distributed_tx/src/test/java/DistributedTxTest.java
package distributed_tx;

import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.BeforeEach;
import static org.junit.jupiter.api.Assertions.*;
import java.util.*;
import java.util.concurrent.*;

// This test suite assumes the existence of the following classes/interfaces in your main source:
// - DistributedTransactionCoordinator with methods:
//       TransactionStatus executeTransaction(String transactionId, List<BranchService> branches)
//       void simulateCrashDuringCommit(String transactionId, List<BranchService> branches)
//       TransactionStatus recoverPendingTransactions(String transactionId)
// - BranchService interface with methods:
//       boolean prepare(String transactionId)
//       void commit(String transactionId)
//       void rollback(String transactionId)
// - TransactionStatus enum with values: COMMITTED and ABORTED

public class DistributedTxTest {

    private DistributedTransactionCoordinator coordinator;

    @BeforeEach
    public void setup() {
        coordinator = new DistributedTransactionCoordinator();
    }

    // Dummy implementation of BranchService for testing purposes.
    public static class DummyBranchService implements BranchService {
        private final boolean failOnPrepare;
        private final long prepareDelayMillis;

        public DummyBranchService(boolean failOnPrepare, long prepareDelayMillis) {
            this.failOnPrepare = failOnPrepare;
            this.prepareDelayMillis = prepareDelayMillis;
        }

        @Override
        public boolean prepare(String transactionId) {
            if (prepareDelayMillis > 0) {
                try {
                    Thread.sleep(prepareDelayMillis);
                } catch (InterruptedException e) {
                    Thread.currentThread().interrupt();
                }
            }
            return !failOnPrepare;
        }

        @Override
        public void commit(String transactionId) {
            // Simulated commit logic.
        }

        @Override
        public void rollback(String transactionId) {
            // Simulated rollback logic.
        }
    }

    // Test a successful transaction where all branches prepare and commit.
    @Test
    public void testSuccessfulTransaction() {
        String transactionId = "tx_success";
        List<BranchService> branches = new ArrayList<>();
        branches.add(new DummyBranchService(false, 0));
        branches.add(new DummyBranchService(false, 0));

        TransactionStatus status = coordinator.executeTransaction(transactionId, branches);
        assertEquals(TransactionStatus.COMMITTED, status, "Expected transaction to be committed");
    }

    // Test a transaction that should abort when one branch fails during the prepare phase.
    @Test
    public void testAbortTransaction() {
        String transactionId = "tx_abort";
        List<BranchService> branches = new ArrayList<>();
        branches.add(new DummyBranchService(false, 0));
        // This branch will simulate a preparation failure.
        branches.add(new DummyBranchService(true, 0));

        TransactionStatus status = coordinator.executeTransaction(transactionId, branches);
        assertEquals(TransactionStatus.ABORTED, status, "Expected transaction to be aborted due to a branch failure");
    }

    // Test a transaction that times out due to one branch's delayed response.
    @Test
    public void testTimeoutTransaction() {
        String transactionId = "tx_timeout";
        List<BranchService> branches = new ArrayList<>();
        // Assuming the coordinator's timeout is less than 1500ms.
        branches.add(new DummyBranchService(false, 1500));
        branches.add(new DummyBranchService(false, 0));

        TransactionStatus status = coordinator.executeTransaction(transactionId, branches);
        // Expect the coordinator to abort if a branch does not respond in time.
        assertEquals(TransactionStatus.ABORTED, status, "Expected transaction to be aborted due to timeout");
    }

    // Test multiple concurrent transactions to ensure proper isolation and synchronization.
    @Test
    public void testConcurrentTransactions() throws InterruptedException, ExecutionException {
        ExecutorService executor = Executors.newFixedThreadPool(5);
        List<Callable<TransactionStatus>> tasks = new ArrayList<>();

        for (int i = 0; i < 10; i++) {
            final String txId = "tx_concurrent_" + i;
            tasks.add(() -> {
                List<BranchService> branches = new ArrayList<>();
                branches.add(new DummyBranchService(false, 0));
                branches.add(new DummyBranchService(false, 0));
                return coordinator.executeTransaction(txId, branches);
            });
        }

        List<Future<TransactionStatus>> futures = executor.invokeAll(tasks);
        for (Future<TransactionStatus> future : futures) {
            TransactionStatus status = future.get();
            assertEquals(TransactionStatus.COMMITTED, status, "Expected concurrent transaction to be committed");
        }
        executor.shutdown();
    }

    // Test the recovery mechanism to handle unfinished transactions after a simulated coordinator crash.
    @Test
    public void testRecovery() {
        String transactionId = "tx_recovery";
        List<BranchService> branches = new ArrayList<>();
        // Use branches with moderate delay to simulate ongoing transaction state.
        branches.add(new DummyBranchService(false, 500));
        branches.add(new DummyBranchService(false, 500));

        // Simulate a crash during the commit phase.
        coordinator.simulateCrashDuringCommit(transactionId, branches);

        // Invoke the recovery mechanism to complete the pending transaction.
        TransactionStatus status = coordinator.recoverPendingTransactions(transactionId);
        assertEquals(TransactionStatus.COMMITTED, status, "Expected recovered transaction to be committed");
    }
}

File: cpp/exercises/practice/unit_test_14211e63_e91c_44b4_8041_ec9b282af1d0/distributed_tx/distributed_tx_test.cpp
#include <chrono>
#include <thread>
#include "catch.hpp"
#include "distributed_tx.h"

class MockParticipant : public IParticipant {
public:
    MockParticipant(bool willSucceed = true, bool shouldTimeout = false) 
        : willSucceed_(willSucceed), shouldTimeout_(shouldTimeout) {}

    PrepareResult prepare() override {
        if (shouldTimeout_) {
            std::this_thread::sleep_for(std::chrono::seconds(5));
        }
        prepared_ = willSucceed_;
        return willSucceed_ ? PrepareResult::READY : PrepareResult::ABORT;
    }

    bool commit() override {
        if (!prepared_) return false;
        if (shouldTimeout_) {
            std::this_thread::sleep_for(std::chrono::seconds(5));
        }
        committed_ = willSucceed_;
        return willSucceed_;
    }

    bool rollback() override {
        if (shouldTimeout_) {
            std::this_thread::sleep_for(std::chrono::seconds(5));
        }
        prepared_ = false;
        committed_ = false;
        return true;
    }

    bool isCommitted() const { return committed_; }
    bool isPrepared() const { return prepared_; }

private:
    bool willSucceed_;
    bool shouldTimeout_;
    bool prepared_ = false;
    bool committed_ = false;
};

TEST_CASE("Single participant successful transaction") {
    TransactionCoordinator coordinator;
    auto participant = std::make_shared<MockParticipant>(true);
    std::vector<std::shared_ptr<IParticipant>> participants = {participant};
    
    REQUIRE(coordinator.executeTransaction(participants) == TransactionResult::COMMITTED);
    REQUIRE(participant->isCommitted());
}

TEST_CASE("Single participant failed preparation") {
    TransactionCoordinator coordinator;
    auto participant = std::make_shared<MockParticipant>(false);
    std::vector<std::shared_ptr<IParticipant>> participants = {participant};
    
    REQUIRE(coordinator.executeTransaction(participants) == TransactionResult::ABORTED);
    REQUIRE_FALSE(participant->isCommitted());
}

TEST_CASE("Multiple participants successful transaction") {
    TransactionCoordinator coordinator;
    auto p1 = std::make_shared<MockParticipant>(true);
    auto p2 = std::make_shared<MockParticipant>(true);
    auto p3 = std::make_shared<MockParticipant>(true);
    std::vector<std::shared_ptr<IParticipant>> participants = {p1, p2, p3};
    
    REQUIRE(coordinator.executeTransaction(participants) == TransactionResult::COMMITTED);
    REQUIRE(p1->isCommitted());
    REQUIRE(p2->isCommitted());
    REQUIRE(p3->isCommitted());
}

TEST_CASE("Multiple participants with one failure") {
    TransactionCoordinator coordinator;
    auto p1 = std::make_shared<MockParticipant>(true);
    auto p2 = std::make_shared<MockParticipant>(false);
    auto p3 = std::make_shared<MockParticipant>(true);
    std::vector<std::shared_ptr<IParticipant>> participants = {p1, p2, p3};
    
    REQUIRE(coordinator.executeTransaction(participants) == TransactionResult::ABORTED);
    REQUIRE_FALSE(p1->isCommitted());
    REQUIRE_FALSE(p2->isCommitted());
    REQUIRE_FALSE(p3->isCommitted());
}

TEST_CASE("Participant timeout during prepare phase") {
    TransactionCoordinator coordinator;
    auto p1 = std::make_shared<MockParticipant>(true);
    auto p2 = std::make_shared<MockParticipant>(true, true);  // Will timeout
    std::vector<std::shared_ptr<IParticipant>> participants = {p1, p2};
    
    REQUIRE(coordinator.executeTransaction(participants) == TransactionResult::TIMEOUT);
    REQUIRE_FALSE(p1->isCommitted());
    REQUIRE_FALSE(p2->isCommitted());
}

TEST_CASE("Concurrent transactions") {
    TransactionCoordinator coordinator;
    std::vector<std::future<TransactionResult>> futures;
    
    for (int i = 0; i < 5; ++i) {
        futures.push_back(std::async(std::launch::async, [&coordinator]() {
            auto p1 = std::make_shared<MockParticipant>(true);
            auto p2 = std::make_shared<MockParticipant>(true);
            std::vector<std::shared_ptr<IParticipant>> participants = {p1, p2};
            return coordinator.executeTransaction(participants);
        }));
    }
    
    for (auto& future : futures) {
        REQUIRE(future.get() == TransactionResult::COMMITTED);
    }
}

TEST_CASE("Empty participant list") {
    TransactionCoordinator coordinator;
    std::vector<std::shared_ptr<IParticipant>> participants;
    
    REQUIRE(coordinator.executeTransaction(participants) == TransactionResult::INVALID);
}

TEST_CASE("Null participant") {
    TransactionCoordinator coordinator;
    std::vector<std::shared_ptr<IParticipant>> participants = {nullptr};
    
    REQUIRE(coordinator.executeTransaction(participants) == TransactionResult::INVALID);
}

TEST_CASE("Maximum participants exceeded") {
    TransactionCoordinator coordinator;
    std::vector<std::shared_ptr<IParticipant>> participants;
    
    for (int i = 0; i < 1000; ++i) {
        participants.push_back(std::make_shared<MockParticipant>(true));
    }
    
    REQUIRE(coordinator.executeTransaction(participants) == TransactionResult::INVALID);
}